### DISCUSSION TITLE
how to use 'useSubmit' for handling multiple actions with one router?

### DISCUSSION THREAD

**Discussion by BigSillyTiger** [remix-run/remix#7565] [Q&A]
Hi everyone:

I have read this(https://reactrouter.com/en/main/route/action#handling-multiple-actions-per-route), but how can I use 'useSubmit' for achieving the same result?

my situation is like: I retrieve data from db by loader, display data with a table, and an action for inserting more data to db
because I am using react-hook-form for form validation, my code is like below: 

`const onSubmit = async (e: FormEvent) => {
        const isValid = await trigger(); // trigger from react-hook-form -> useForm
        e.preventDefault();
        if (isValid) {
            const values = getValues();
            submit(values, { action: "/clients", method: "POST" });
        }
    };
...
<Form onSubmit={onSubmit}>
...
`
my troubles:
1. when I simply emitted a delete request, after the execution, I had to manually update the table, 
2. for updating the data through another form with validation of react-hook-form, what shall I do?

**Comment by kiliman** [comment]
I'm not quite sure what you are asking. Remix should revalidate your routes after a submit. The loaders should return the updated data. 

Do you have an example repo showing your issue?

**Reply by BigSillyTiger** [reply]
Thanks for the reply, here is my repo: [repo](https://github.com/BigSillyTiger/Alax-Tax-Frontend/blob/clients-menu/src/pages/clients/clients.tsx)
my /src/page/client component is used to display a client list with a table, along with adding new clients, deleting clients, and updating (I have not implemented this one yet) operations, and maybe more data operations in the future. 

1. [add-new-client](https://github.com/BigSillyTiger/Alax-Tax-Frontend/blob/clients-menu/src/pages/clients/modals/mAddNewClient.tsx) in this file, I have a form for adding a new client to my database, I am using react-hook-form for validation so I introduced useSubmit from react-router, till now everything is working alright. Here the action is {method: 'POST', action: '/clients'}
2. [display-one-client](https://github.com/BigSillyTiger/Alax-Tax-Frontend/blob/clients-menu/src/pages/clients/modals/mClient.tsx) in this file, I introduced a button for open a new [modal](https://github.com/BigSillyTiger/Alax-Tax-Frontend/blob/clients-menu/src/pages/clients/modals/mDelete.tsx) to confirm the delete operation. if YES, then I simply emit a delete request to the backend and delete the client from my database. In this case, however, **the table will not automatically rerender since I did not use the action to delete the client(because there's no form), which will not trigger the loader to re-fetch new data from the backend.** my original idea was to use 'useSubmit' with 'onClick' property to fire the {method: 'DELETE', action: '/clients'}, but I don't know how to separate this from the previous {method: 'POST', action: '/clients'} in the action function.
3. for the updating operation in the future, I will introduce {method: 'PUT', action: '/clients'} for updating the client info in my database, probably still using 'useSubmit' then the issue in step 2 will come again.

I am a totally rookie developer and still learning and progressing, so basically everything mentioned by other people is all 'brandly new' to me... and I just learned some basics of react-router, and did not pay any attention to Remix which I'll try right now, but is this just for validation? Will the multiple actions with one router issue be fixed through it?

PS: [andling-multiple-actions-per-route](https://reactrouter.com/en/main/route/action#handling-multiple-actions-per-route) this does not fit my case ...

again, thank you for helping me

**Comment by BigSillyTiger** [comment]
I just found out that I could use 
`
if ("POST" === request.method) 
...
else if ("DELETE" === request.method)`
 to make my [actions](https://github.com/BigSillyTiger/Alax-Tax-Frontend/blob/clients-menu/src/pages/clients/dataAccess.ts) function distinguish different methods, this could fix my current issue, but then a new issue comes: what if I have more than one method of the same type? like 2 POST or 2 DELETE, but they are of different purposes?

---
or, I can add an pair of extra key value in useSubmit/submit for action function like: 
`submit({..., {type: 'f1' } }, { method: "DELETE", action: "/clients" });`
`if ("DELETE" === request.method && data.get("type") === 'f1')`
